### PR TITLE
netcheck: add hairpinning detection

### PR DESCRIPTION
**Pull request by bradfitz**
(I want to refactor this code a bit later. This change is getting near pushing me over the line to do that.)

### CODE REVIEW

File: netcheck/netcheck.go [hunk 1]
```diff
@@ -24,10 +26,13 @@ type Report struct {
 	UDP                   bool                     // UDP works
 	IPv6                  bool                     // IPv6 works
 	MappingVariesByDestIP opt.Bool                 // for IPv4
+	HairPinning           opt.Bool                 // for IPv4
 	DERPLatency           map[string]time.Duration // keyed by STUN host:port
 }
 
 func GetReport(ctx context.Context, logf logger.Logf) (*Report, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
Comment (crawshaw) [inline netcheck/netcheck.go]: Maybe add a comment explaining that this ctx is to ensure closeOnCtx is called on return.

**Comment by bradfitz** [comment]
So, this "works" ... it reports false on VMs inside my NAT, and true on my Linux gateway doing that NAT.

So I guess we could use this to rule out connecting to a peer's global IPv4 address if both nodes have the same global IPv4 and either has HairPinning==false?

File: netcheck/netcheck.go (hunk 1 above)
Comment (bradfitz) [reply netcheck/netcheck.go]: Done.

**Comment by bradfitz** [comment]
Merged via git.

**Pull request closed without merging**